probe_src/probe_py/analysis.py

```diff
@@ -189,6 +208,127 @@ def add_edges(edges:list[tuple[Node, Node]], label:EdgeLabels) -> None:
+            elif access_mode == 2:
+                console.print(f"Found file {op.path.path} with access mode O_RDWR", style="red")
+            else:
+                raise Exception("unknown access mode")
+        elif isinstance(op, CloneOp):
+            if op.task_type == TaskType.TASK_PID:
+                if edge[0][0] != edge[1][0]:
+                    target_nodes[op.task_id].append(edge[1])
+                    continue
+            elif op.task_type == TaskType.TASK_PTHREAD:
+                if edge[0][2] != edge[1][2]:
+                    target_nodes[op.task_id].append(edge[1])
+                    continue
+            if op.task_type != TaskType.TASK_PTHREAD and op.task_type != TaskType.TASK_ISO_C_THREAD:
+                
+                processNode1 = ProcessNode(pid = pid, cmd=" ".join(cmd_map[pid]))
+                processNode2 = ProcessNode(pid = op.task_id, cmd=" ".join(cmd_map[op.task_id]))
+                   
+                dataflow_graph.add_edge(processNode1, processNode2)
+            target_nodes[op.task_id] = list()
+        elif isinstance(op, WaitOp) and op.options == 0:
+            for node in target_nodes[op.task_id]:
+                traverse_hb_for_dfgraph(process_tree_prov_log, node, traversed, dataflow_graph, file_version_map, shared_files, cmd_map, cmd)
+                traversed.append(node[2])
+        # return back to the WaitOp of the parent process
+        if isinstance(next_op, WaitOp):
+            if next_op.task_id == starting_pid or next_op.task_id == starting_op.pthread_id:
+                return
+    return
+
+def list_edges_from_start_node(graph: nx.DiGraph, start_node: Node) -> list[EdgeType]:
+    all_edges = list(graph.edges())
+    start_index = next(i for i, edge in enumerate(all_edges) if edge[0] == start_node)
+    ordered_edges = all_edges[start_index:] + all_edges[:start_index] 
+    return ordered_edges
+
+def provlog_to_dataflow_graph(process_tree_prov_log: ProvLog) -> nx.DiGraph:
+    dataflow_graph = nx.DiGraph()
+    file_version_map: Dict[str, int] = {}
+    process_graph = provlog_to_digraph(process_tree_prov_log)
+    root_node = [n for n in process_graph.nodes() if process_graph.out_degree(n) > 0 and process_graph.in_degree(n) == 0][0]
+    traversed: list[int] = []
+    cmd:list[str] = []
+    cmd_map:Dict[int, list[str]] = {}
+    for edge in list(nx.edges(process_graph))[::-1]:
+        pid, exec_epoch_no, tid, op_index = edge[0]
+        op = prov_log_get_node(process_tree_prov_log, pid, exec_epoch_no, tid, op_index).data
+        if isinstance(op, OpenOp):
+            file = op.path.path
+            if file not in cmd and file!="/dev/tty":
+                cmd.insert(0, file)
+        elif isinstance(op, InitExecEpochOp):
+            program_name = op.program_name
+            cmd.insert(0, program_name)
+            if pid == tid and exec_epoch_no == 0:
+                cmd_map[tid] = cmd
+                cmd = []
```

Comment:
I don't think this is valid. One program might have two processes with the exact same command. Imagine a script like `cat a.txt ; cat a.txt`. The identity of a process should solely be determined by the PID which would be different for each cat even though the string-command is the same. The program name should be used solely for displaying purposes. I will capture the rest of the arguments in in the `ExecOp` in a later version of libprobe.